plugins/state_history_plugin/state_history_plugin.cpp

```diff
@@ -338,61 +338,46 @@ void state_history_plugin::plugin_initialize(const variables_map& options) {
-         if( first_state_block > 0 )
-            first_available_block = std::min( first_available_block, first_state_block );
-      }
-      if(finality_data_log) {
-         auto first_state_block = finality_data_log->block_range().first;
-         if( first_state_block > 0 )
-            first_available_block = std::min( first_available_block, first_state_block );
-      }
-      fc_ilog(_log, "First available block for SHiP ${b}", ("b", first_available_block));
-      listen();
-      thread_pool.start(1, [](const fc::exception& e) {
-         fc_elog( _log, "Exception in SHiP thread pool, exiting: ${e}", ("e", e.to_detail_string()) );
-         app().quit();
-      });
-   } catch(std::exception& ex) {
-      appbase::app().quit();
+   uint32_t block_num = chain.head().block_num();
+   if( block_num > 0 && chain_state_log && chain_state_log->empty() ) {
+      fc_ilog( _log, "Storing initial state on startup, this can take a considerable amount of time" );
+      store_chain_state( chain.head().id(), chain.head().header().previous, block_num );
+      fc_ilog( _log, "Done storing initial state on startup" );
+   }
+   first_available_block = chain.earliest_available_block_num();
+   if(trace_log) {
+      auto first_trace_block = trace_log->block_range().first;
+      if( first_trace_block > 0 )
+         first_available_block = std::min( first_available_block, first_trace_block );
+   }
+   if(chain_state_log) {
+      auto first_state_block = chain_state_log->block_range().first;
+      if( first_state_block > 0 )
+         first_available_block = std::min( first_available_block, first_state_block );
    }
+   if(finality_data_log) {
+      auto first_state_block = finality_data_log->block_range().first;
+      if( first_state_block > 0 )
+         first_available_block = std::min( first_available_block, first_state_block );
+   }
+   fc_ilog(_log, "First available block for SHiP ${b}", ("b", first_available_block));
+   listen();
+   thread_pool.start(1, [](const fc::exception& e) {
+      fc_elog( _log, "Exception in SHiP thread pool, exiting: ${e}", ("e", e.to_detail_string()) );
+      app().quit();
+   });
 }
 
 void state_history_plugin::plugin_startup() {
    my->plugin_startup();
 }
 
 void state_history_plugin_impl::plugin_shutdown() {
-   applied_transaction_connection.reset();
-   accepted_block_connection.reset();
-   block_start_connection.reset();
+   fc_dlog(_log, "stopping");
    thread_pool.stop();
-
-   // This is a temporary fix until https://github.com/AntelopeIO/spring/issues/842 can be worked.
-   // Drain the io_service of anything that could be referencing state_history_plugin
-   if (app().executor().get_io_service().stopped()) {
-      app().executor().get_io_service().restart();
-      while (app().executor().get_io_service().poll())
-         ;
-      // clear priority queue of anything pushed by poll(), see application_base exec()
-      app().executor().clear();
-   }
+   fc_dlog(_log, "exit shutdown");
```

Comment:
This is inconsistent with the other changes where the 'exit shutdown' is printed at info level, for example
https://github.com/AntelopeIO/spring/blob/2d8b042e111cb04538af8fb1abe4181e6583dbc8/plugins/resource_monitor_plugin/resource_monitor_plugin.cpp#L119
and
https://github.com/AntelopeIO/spring/blob/2d8b042e111cb04538af8fb1abe4181e6583dbc8/plugins/net_plugin/net_plugin.cpp#L4432